src/Microsoft.AspNetCore.Mvc.Razor.Host/Internal/ViewComponentTagHelperChunkVisitor.cs

```diff
@@ -0,0 +1,189 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Chunks;
+using Microsoft.AspNetCore.Razor.CodeGenerators;
+using Microsoft.AspNetCore.Razor.CodeGenerators.Visitors;
+using Microsoft.AspNetCore.Razor.Compilation.TagHelpers;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Host.Internal
+{
+    public class ViewComponentTagHelperChunkVisitor : CodeVisitor<CSharpCodeWriter>
+    {
+        private readonly GeneratedViewComponentTagHelperContext _context;
+        private readonly HashSet<string> _writtenViewComponents;
+
+        private const string ViewComponentTagHelperVariable = "_viewComponentHelper";
+        private const string _viewContextVariable = "ViewContext";
+
+        public ViewComponentTagHelperChunkVisitor(CSharpCodeWriter writer, CodeGeneratorContext context) :
+            base(writer, context)
+        {
+            _context = new GeneratedViewComponentTagHelperContext();
+            _writtenViewComponents = new HashSet<string>(StringComparer.Ordinal);
+        }
+
+        public override void Accept(Chunk chunk)
+        {
+            if (chunk == null)
+            {
+                throw new ArgumentNullException(nameof(chunk));
+            }
+
+            var tagHelperChunk = chunk as TagHelperChunk;
+            if (tagHelperChunk != null) 
+            {
+                Visit(tagHelperChunk);
+            }
+
+            var parentChunk = chunk as ParentChunk;
+            if (parentChunk != null)
+            {
+                Accept(parentChunk.Children);
+            }
+        }
+
+        protected override void Visit(TagHelperChunk chunk)
+        {
+            var viewComponentDescriptors = chunk.Descriptors
```

Comment:
Would it make sense to `TryGetValue` the `ViewComponentNameKey` and skip the descriptor if the property doesn't exist? Basically avoids looking up the key twice in the dictionary + fewer Linq calls.